crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,237 @@
+//! Different formatting functions.
+
+use std::time::Duration;
+
+/// Returns value with `n` digits after floating point where `n` is `precision`.
+/// Standard rounding rules apply (if `n+1`th digit >= 5, round up).
+///
+/// If rounding the `value` will have no effect (e.g., it's infinite or
+/// NaN), returns `value` unchanged.
+///
+/// # Examples
+///
+/// ```
+/// # use typst_utils::format::round_with_precision;
+/// let rounded = round_with_precision(-0.56553, 2);
+/// assert_eq!(-0.57, rounded);
+/// ```
+pub fn round_with_precision(value: f64, precision: u8) -> f64 {
+    // Don't attempt to round the float if that wouldn't have any effect. This
+    // includes infinite or NaN values, as well as integer values
+    // with a filled mantissa (which can't have a fractional part). Rounding
+    // with a precision larger than the amount of digits that can be
+    // effectively represented would also be a no-op.
+    if value.is_infinite()
+        || value.is_nan()
+        || value.abs() >= (1_i64 << f64::MANTISSA_DIGITS) as f64
+        || precision as u32 >= f64::DIGITS
+    {
+        return value;
+    }
+    let offset = 10_f64.powi(precision.into());
+    assert!((value * offset).is_finite(), "{value} * {offset} is not finite!");
+    (value * offset).round() / offset
+}
+
+/// Returns `(days, hours, minutes, seconds, milliseconds, microseconds)`.
+fn get_duration_parts(duration: &Duration) -> (u16, u8, u8, u8, u16, u16) {
```

Comment:
This can overflow with reasonably large days, so we should probably update this signature, and change this in other places taking days too (like `format_dhms` and others):

```suggestion
fn get_duration_parts(duration: &Duration) -> (u64, u8, u8, u8, u16, u16) {
```

Reply:
I thought we're trying to make the app faster, not slower. :(

How can you wait for 180 years to compile?

I can only see a problem of a misuse of the function by feeding in a huge duration on purpose (not possible in the CLI) through the public functions. But that would be a misuse and can be caught with a panic.

So in any case, if the function is used as intended, the 16 bits is more than enough for a days counter.

Reply:
Yeah I doubt this would be triggered naturally, just wanted to have a type matching Duration's `from_days` for full accuracy. Doesn't hurt to have some extra caution here, especially because the function is public.

Reply:
This also would require changing `format_dhms`'s signature. Do I still proceed?

Reply:
Yes